src/mds/Server.cc

```diff
@@ -767,6 +781,17 @@ void Server::find_idle_sessions()
       dout(20) << "oldest stale session is " << session->info.inst << " and sufficiently new (" 
 	       << session->last_cap_renew << ")" << dendl;
       break;
+    } else if (session->eviction_deferred and cutoff -
+		session->last_cap_renew <  2 *
+		mds->mdsmap->get_session_autoclose())
+      continue;
+
+    if (not mds->locker->is_revoking_any_caps_from((session)->get_client()) and
+	not session->eviction_deferred) {
```

Comment:
is_revoking_any_caps_from() should always return false for stale session

Reply:
Copying the modified code here, so that we can see with a gdb session if `is_revoking_any_caps_from()`  returns true due to branching and to report the branch taken in the logs -

```
    if (not mds->locker->is_revoking_any_caps_from((session)->get_client())) {
        if (session->eviction_deferred) {
          dout(1) << "is_revoking_any_caps_from returns false" << dendl;
          dout(20) << "deferring autoclose for client " << session->
                  get_client().v << " since it holds no caps" << dendl;
          session->eviction_deferred = true;
          continue;
        }
    } else
      dout(1) << "is_revoking_any_caps_from returns true" << dendl;
```
gdb session -

```
Breakpoint 1, Server::find_idle_sessions (this=0x555ae7eaabd0) at /home/centos/repos/ceph/src/mds/Server.cc:789
789	    if (not mds->locker->is_revoking_any_caps_from((session)->get_client())) {
(gdb) n
798	      dout(1) << "is_revoking_any_caps_from returns true" << dendl;
(gdb) n
800	    to_evict.push_back(session);
```

grep-ing out/mds.a.log; take a look at the last line -

```
$ grep -rnI true out/mds.a.log 
5:2018-07-02 08:44:44.005 7f7e0583d700 10 set_mon_vals mds_debug_auth_pins = true
7:2018-07-02 08:44:44.005 7f7e0583d700 10 set_mon_vals mds_debug_frag = true
9:2018-07-02 08:44:44.005 7f7e0583d700 10 set_mon_vals mds_debug_scatterstat = true
11:2018-07-02 08:44:44.005 7f7e0583d700 10 set_mon_vals mds_debug_subtrees = true
13:2018-07-02 08:44:44.005 7f7e0583d700 10 set_mon_vals mds_verify_scatter = true
108:2018-07-02 08:44:44.032 7f7e0403a700 20 set_mon_vals mds_debug_auth_pins = true (no change)
110:2018-07-02 08:44:44.032 7f7e0403a700 20 set_mon_vals mds_debug_frag = true (no change)
112:2018-07-02 08:44:44.032 7f7e0403a700 20 set_mon_vals mds_debug_scatterstat = true (no change)
114:2018-07-02 08:44:44.032 7f7e0403a700 20 set_mon_vals mds_debug_subtrees = true (no change)
117:2018-07-02 08:44:44.032 7f7e0403a700 20 set_mon_vals mds_verify_scatter = true (no change)
261:2018-07-02 08:44:51.377 7f7e0583d700 20 log_channel(cluster) update_config log_to_monitors {default=true} log_to_syslog {default=false} log_channels {audit=local0,default=daemon} log_prios {default=info}
262:2018-07-02 08:44:51.377 7f7e0583d700 10 log_channel(cluster) update_config to_monitors: true to_syslog: false syslog_facility: daemon prio: info to_graylog: false graylog_host: 127.0.0.1 graylog_port: 12201)
263:2018-07-02 08:44:51.377 7f7e0583d700 10 mds.0.0 update_log_config log_to_monitors {default=true}
3616:2018-07-02 08:48:06.506 7f7e03038700  1 mds.0.server is_revoking_any_caps_from returns true
```

Reply:
above session is in open state, why did you find it in stale session list

Reply:
must be bug of is_revoking_any_caps_from(). I checked code that modifies revoking_caps_by_client, there is no code that erases client from revoking_caps_by_client (when revoking_caps_by_client[client] is empty)

Reply:
Yeah, it could be a bug...
   

Modified `is_revoking_any_caps_from()` for more debug messages -
```
bool Locker::is_revoking_any_caps_from(client_t client)
{
  auto rlist = revoking_caps_by_client;
      
  if (rlist.empty())
    dout(0) << "token, empty!" << dendl;
  else 
    dout(0) << "token, not empty!" << dendl;
  for (auto r : rlist)
    dout(0) << "token, client id = " << r.first.v << dendl;
  
  return true ? rlist.find(client) != rlist.end() : false;
}
```

Steps after setting up vstart cluster and shortening of `session_timeout` and `session_autoclose` -
```
sudo ./bin/ceph-fuse /mnt/ceph-fuse
sudo ./bin/ceph-fuse /mnt/ceph-fuse2
./bin/ceph-fuse tell mds.a client ls  # report 2 clients
cp ~/dummyfs/* /mnt/cephfs    # copying some non-sense files and dirs
sleep $((SESSION_TIMEOUT + SESSION_AUTOCLOSE))
./bin/ceph-fuse tell mds.a client ls  #reports only 1 client
```
I see that `revoking_caps_by_client` is not empty, so `true` is returned and clients get evicted.

```
$ grep -rnI token out/mds.a.log
3141:2018-07-04 10:22:31.714 7f09493d9700  0 mds.0.locker token, not empty!
3142:2018-07-04 10:22:31.714 7f09493d9700  0 mds.0.locker token, client id = 4151
3143:2018-07-04 10:22:31.714 7f09493d9700  0 mds.0.locker token, client id = 4153
3197:2018-07-04 10:22:31.714 7f09493d9700  0 mds.0.locker token, not empty!
3198:2018-07-04 10:22:31.714 7f09493d9700  0 mds.0.locker token, client id = 4151
3199:2018-07-04 10:22:31.714 7f09493d9700  0 mds.0.locker token, client id = 4153
3599:2018-07-04 10:23:06.714 7f09493d9700  0 mds.0.locker token, not empty!
3600:2018-07-04 10:23:06.714 7f09493d9700  0 mds.0.locker token, client id = 4151
3601:2018-07-04 10:23:06.714 7f09493d9700  0 mds.0.locker token, client id = 4153
3697:2018-07-04 10:23:31.313 7f09493d9700  0 mds.0.locker token, not empty!
3698:2018-07-04 10:23:31.313 7f09493d9700  0 mds.0.locker token, client id = 4151
3699:2018-07-04 10:23:31.313 7f09493d9700  0 mds.0.locker token, client id = 4153
3765:2018-07-04 10:23:48.800 7f09493d9700  0 mds.0.locker token, not empty!
3766:2018-07-04 10:23:48.800 7f09493d9700  0 mds.0.locker token, client id = 4151
3767:2018-07-04 10:23:48.800 7f09493d9700  0 mds.0.locker token, client id = 4153
$
```

I thought there might be a corner case (e.g. when list is empty), but it is not so.

Reply:
I mean

bool Locker::is_revoking_any_caps_from(client_t client)
{
  auto it = revoking_caps_by_client.find(client);
  if (it == revoking_caps_by_client.end())
    return false;
  return !it->second.empty();      
}